article.md

```diff
@@ -0,0 +1,333 @@
+`Payload` will be the main structure that we'll be constructing and signing. For the sake of simplicity, it only has one field but the EIP-712 standard allows for any type of structure.
+
+## Creating a connection
+On the surface the concept might be simple, we get the Ethereum address as an input and use the caller's address as Starknet address, we call `save_connected_addresses` function and we're done, right? Well not really, first we have to make sure that the Ethereum address belongs to the person calling this function. That's the hard part, we need to recreate the message signed on the Ethereum side, and then by using the signature given as an input we need to confirm that the extracted address matches our input. That what `assert_valid_eth_signature` function is doing. By calling `get_caller_address()` and using it as our Starknet address we ensure that it indeed belongs to the user and we will use it when reconstructing a signed message.
+```cairo
+@external
+func add_connection{
+    syscall_ptr : felt*,
+    pedersen_ptr : HashBuiltin*,
+    range_check_ptr,
+    bitwise_ptr : BitwiseBuiltin*
+}(eth_address : felt, starknet_address : felt, domain_hash: Uint256, signature_len: felt, signature: felt*){
+    assert_valid_eth_signature(eth_address, starknet_address, domain_hash, signature_len, signature);
+    save_connected_addresses(eth_address, starknet_address);
+    return ();
+}
+```
+
+The hardest part for us is getting the hash of our message. Once we have that, we can use a built-in function `verify_eth_signature_uint256` that will do all the heavy lifting, dealing with complicated math. 
+So we will have to write the `get_hash` function for ourselves.
+```cairo
+func assert_valid_eth_signature{
+        syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr,
+        bitwise_ptr : BitwiseBuiltin*
+}(eth_address: felt, starknet_address: felt, domain_hash: Uint256, signature_len: felt, signature: felt*){
+    alloc_locals;
+    with_attr error_message(
+        "Invalid signature length. Signature should have exactly 5 elements."
+    ){
+        assert signature_len = 5;
+    }
+
+    let v = signature[0];
+    let r = Uint256(signature[2], signature[1]);
+    let s = Uint256(signature[4], signature[3]);
+    
+    let (local keccak_ptr_start) = alloc();
+    let keccak_ptr = keccak_ptr_start;
+
+    let (hash_uint) = get_hash{keccak_ptr=keccak_ptr}(starknet_address, domain_hash);
+   
+    verify_eth_signature_uint256{keccak_ptr=keccak_ptr}(hash_uint, r, s, v, eth_address);
+
+    finalize_keccak(keccak_ptr_start=keccak_ptr_start, keccak_ptr_end=keccak_ptr);
+
+    return ();
+}
+```
+
+## Signing with an Ethereum wallet
```

Comment:
This would should be before any code, to show how someone can generate signatures in the first place.